### ISSUE TITLE
Error type related to the struct contaning global buffer reference

### ISSUE THREAD

**Issue by NoSW** [KhronosGroup/glslang#3188]
Part of glsl code (fragment shader):
```glsl
#version 450 core
#extension GL_EXT_scalar_block_layout : require
#extension GL_EXT_buffer_reference : enable
#extension GL_EXT_scalar_block_layout : enable

layout(buffer_reference, std430) buffer MeshVertexPositions {
  float data[];
};

layout(buffer_reference, std430) buffer MeshVertexNormals {
  float data[];
};

layout(buffer_reference, std430) buffer MeshVertexTangents {
  float data[];
};

layout(buffer_reference, std430) buffer MeshVertexUV0s {
  vec2 data[];
};

layout(buffer_reference, std430) buffer MeshVertexUV1s {
  vec2 data[];
};

layout(buffer_reference, std430) buffer MeshIndices {
  uint data[];
};

struct Mesh {
  MeshVertexPositions positions;
  MeshVertexNormals normals;
  MeshVertexTangents tangents;
  MeshVertexUV0s uv0s;
  MeshVertexUV1s uv1s;
  MeshIndices indices;
  uint clusterOffset;
  uint clusterCount;
  float aabb[6];
};

layout(set = 0, binding = 7) readonly buffer PerPass_meshes {
  Mesh data[];
} perPass_meshes;

void doInterpolation(in unit meshId,  in unit triangleId, ...){
    Mesh meshData = perPass_meshes.data[meshId];

    // read data from meshData 
}
```

Part of assembly from spirv-dis:
```
%4057 = OpAccessChain %_ptr_StorageBuffer_Mesh_0 %perPass_meshes %int_0 %4055
       %4058 = OpLoad %Mesh_0 %4057
       %4059 = OpCopyLogical %Mesh %4058
       %8428 = OpCompositeExtract %_ptr_PhysicalStorageBuffer_MeshVertexTangents %4059 0   // Error!
       %8429 = OpCompositeExtract %_ptr_PhysicalStorageBuffer_MeshVertexTangents %4059 1
       %8430 = OpCompositeExtract %_ptr_PhysicalStorageBuffer_MeshVertexTangents %4059 2
       %8431 = OpCompositeExtract %_ptr_PhysicalStorageBuffer_MeshVertexUV1s %4059 3
       %8432 = OpCompositeExtract %_ptr_PhysicalStorageBuffer_MeshVertexUV1s %4059 4
       %8433 = OpCompositeExtract %_ptr_PhysicalStorageBuffer_MeshIndices %4059 5
       %8434 = OpCompositeExtract %uint %4059 6
               OpLine %1 1203 0
       %4063 = OpIAdd %uint %8434 %3911
               OpLine %1 1205 0
```

Validation error when run application:
```
VULKAN ERROR: (UNASSIGNED-CoreValidation-Shader-InconsistentSpirv) Validation Error: [ UNASSIGNED-CoreValidation-Shader-InconsistentSpirv ] 
Object 0: handle = 0x261430f44a0, name = Device NVIDIA GeForce RTX 2060 SUPER, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x6bbb14 | 
SPIR-V module not valid: Result type (OpTypePointer) does not match the type that results from indexing into the composite (OpTypePointer).

  %8428 = OpCompositeExtract %_ptr_PhysicalStorageBuffer_MeshVertexTangents %4059 0
```

**Comment by NoSW** [comment]
A complete fragment shader file is attached in [glslang#3188.zip](https://github.com/KhronosGroup/glslang/files/11224475/glslang.3188.zip). Also attached a assembly code file copied from the RenderDoc(v1.26)  Pipeline State->Edit->Decompile with spirv-dis. The bug above can be found in line 2240 of this assembly file. 



My vulkan app uses glslang source(lastest version 12.1.0).  So, I also tried to reproduce this bug using standalone command line tools , run
```
glslangValidator.exe --target-env vulkan1.3 -H -S frag -o frag.spv  frag.glsl 
```
then run
```
spirv-dis.exe -o frag.ass frag.spv
```

, but it failed. I didn't find the relevant error in frag.ass.

**Comment by arcady-lunarg** [comment]
I did reproduce this but it only happens with the -Os option enabled, which points to a bug in the spirv-opt tool.

**Comment by arcady-lunarg** [comment]
I'm going to try to reduce your test case and once I do that, I will open an issue in the SPIRV-Tools repository. We'll keep this one open too though, since we will need to update the version of SPIRV-Tools against which glslang is built.

**Comment by NoSW** [comment]
In addition to spirv-opt, glslang may also has the bug handling `struct Mesh { ... };`

Running in Debug mode: (version 12.1.0)
```
glslangValidator.exe  -H -g -gV -gVS --target-env vulkan1.3 -S frag -o frag.spv frag.glsl
```
will trigger `assert(debugId[memberType] != 0)`  in glslang\SPIRV\SpvBuilder.cpp:967.

**Comment by arcady-lunarg** [comment]
That's a separate bug, I'll create another issue for that.

**Comment by arcady-lunarg** [comment]
It looks like the fix for the spirv-opt side of the issue has landed, however it still requires a fix on the glslang side as glslang is generating the struct type without the correct decoration here. I will work on the glslang fix and also on updating the version of spirv-tools used by glslang.

**Comment by arcady-lunarg** [comment]
For what it's worth, I think this specific case should be fixed by the version of spirv-tools included with the 12.2 release, however there is still a spec compliance issue with decorations that will be fixed with PR #3229.
